### PR TITLE
feat: Added Dynamic Link Previews

### LINKED ISSUE #1064: [FEATURE][FormTribe 🔥][500 Points] Dynamic Link Previews
### Welcome to the FormTribe

This issue will get you the points mentioned in issue title for the [FormTribe Hackathon](https://formtribe.com)

**Preparation:**
If this is your first time contributing, please read:

- Our [introductory blog post](https://formbricks.com/blog/join-the-formtribe)
- Overview of [Code Guidelines at Formbricks.](https://formbricks.com/docs/contributing/how-we-code)

**Only when your PR follows _all_ of the best practices it will get merged.**

Want more points quick? Complete [Side Quests 🏰 ](https://formbricks.com/formtribe#side-quests)

--------

### Is your feature request related to a problem? Please describe.

Link previews are not optimized:

image

### Describe the solution you'd like

Leverage next.js dynamic link preview feature for better link previews:

image

### Tasks

- Use nextJS [dynamic image generation](https://cruip.com/generate-dynamic-open-graph-and-twitter-images-in-next-js/#:~:text=In%20summary%2C%20it%20involves%20generating,in%20just%20a%20few%20milliseconds!) to create the thumbnail for each survey
- When generating it, make sure that the color of the "Begin" button is set to the brandColor value of the product.
- The "Complete in ~4 Minutes" is static for now, the calculation will be done in a different ticket next week :) We will then update it here, but it shouldn't stop us from shipping 🚢

### Technical Details 
- Utilize the next/head component to insert these tags into the head of the rendered page. You need to add some meta tags like :

- `og:title:` Reflecting the survey's name.
- `og:description:` The "Complete in 4 Minutes"

You can dynamically add these in 

`apps/web/app/s/[surveyId]/page.tsx`

You wont be able to test this on localhost. So you can use something like **ngrok** .

ngrok grants a public URL, mapped to our local server, perfect for real-time testing. Sharing links via this URL on various platforms can help verify the dynamic generation of our meta tags. 

You can test the previews here : https://link-previews.stephanbogner.de/

Please follow [this tutorial](https://cruip.com/generate-dynamic-open-graph-and-twitter-images-in-next-js/#:~:text=In%20summary%2C%20it%20involves%20generating,in%20just%20a%20few%20milliseconds!) for more context on how to get it done :)

### PR DESCRIPTION
**Pull request by anjy7**
## What does this PR do?

Adds Dynamic Link Previews

Fixes #1064 

Screenshot from 2023-10-11 21-07-22
## Type of change

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] Chore (refactoring code, technical debt, workflow improvements)
- [X] Enhancement (small improvements)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change adds a new database migration
- [ ] This change requires a documentation update

## How should this be tested?

- Share survey link

### Required

- [X] Filled out the "How to test" section in this PR
- [X] Read [How we Code at Formbricks](<[https://github.com/formbricks/formbricks/blob/main/CONTRIBUTING.md](https://formbricks.com/docs/contributing/how-we-code)>)
- [X] Self-reviewed my own code
- [ ] Commented on my code in hard-to-understand bits
- [X] Ran `pnpm build`
- [X] Checked for warnings, there are none
- [X] Removed all `console.logs`
- [X] Merged the latest changes from main onto my branch with `git pull origin main`
- [X] My changes don't cause any responsiveness issues

### Appreciated

- [ ] If a UI change was made: Added a screen recording or screenshots to this PR
- [ ] Updated the Formbricks Docs if changes were necessary

### CODE REVIEW

**Comment by jobenjada** [comment]
@anjy7 can you add a screenshot or recording for reference? :)

**Comment by anjy7** [comment]
> @anjy7 can you add a screenshot or recording for reference? :)

Yeah its a work in progress, will add soon

**Comment by jobenjada** [comment]
hey hey!  I tried two different ways to test this and it doesnt work:

image

image


And Im getting this error:
image

Is next/head still the right approach for app dir? 🤔

Could you maybe create a Loom for me to better understand how you are testing this ?

**Comment by anjy7** [comment]
> hey hey! I tried two different ways to test this and it doesnt work:
> 
> image image
> And Im getting this error: image
> 
> Is next/head still the right approach for app dir? 🤔
> 
> Could you maybe create a Loom for me to better understand how you are testing this ?
> 
> Additionally, please resolve the merge conflict :)

sure, I'll look into this and get back to you @jobenjada

**Comment by jobenjada** [comment]
@anjy7 perfect :) I know it'll be Monday morning, but can you maybe have a look tomorrow? :D Would love to include this in our Week 2 Review which I'll record tomorrow afternoon (would be evening in India)

**Comment by anjy7** [comment]
Hey @jobenjada, trying to debug it at presently. And sure I’ll try to do it before evening :)

**Comment by anjy7** [comment]
Hey @jobenjada , I have made the changes, please have a look.
Here's the loom recording as well - https://www.loom.com/share/35345359f50d4e219f804bb47a4d20c8?sid=dd014683-c1da-48b5-893e-15d490faac54

File: apps/web/app/s/[surveyId]/layout.tsx [hunk 1]
```diff
@@ -1,6 +1,24 @@
 import LegalFooter from "@/app/s/[surveyId]/components/LegalFooter";
+import { type Metadata } from "next/types";
 
-export default function SurveyLayout({ children }) {
+export const metadata: Metadata = {
+  title: "Formbricks",
```
Comment (jobenjada) [inline apps/web/app/s/[surveyId]/layout.tsx:5]: You have hard coded the meta data.

I think it makes more sense to create it dynamically using the Survey Title. Is this available on the public facing pages?

File: apps/web/app/s/[surveyId]/layout.tsx [hunk 2]
```diff
@@ -1,6 +1,24 @@
...
+  description: "Open-Source In-Product Survey Platform",
```
Comment (jobenjada) [inline apps/web/app/s/[surveyId]/layout.tsx:6]: Change to "Create your own survey like this with Formbricks' open source survey suite."

File: apps/web/app/s/[surveyId]/page.tsx [hunk 3]
```diff
@@ -25,8 +27,54 @@ interface LinkSurveyPageProps {
...
+  }
+
+  const brandColor = getBrandColorForURL(product.brandColor);
+  const surveyName = getNameForURL(survey.name);
+
+  const ogImgURL = `${WEBAPP_URL}/api/v1/environments/${survey.environmentId}/surveys/${survey.id}/og?brandColor=${brandColor}&name=${surveyName}`;
+
+  return {
+    openGraph: {
+      title: "Formbricks",
```
Comment (jobenjada) [inline apps/web/app/s/[surveyId]/page.tsx:58]: Same here, please fill dynamically

File: apps/web/pages/api/v1/environments/[environmentId]/surveys/[surveyId]/og/index.tsx [hunk 4]
```diff
@@ -0,0 +1,46 @@
...
+// App router includes @vercel/og.
+// No need to install it.
+
+export const runtime = "edge";
+
+export default async function GET(req: NextRequest) {
+  let name = req.nextUrl.searchParams.get("name");
+  let brandColor = req.nextUrl.searchParams.get("brandColor");
+
+  return new ImageResponse(
```
Comment (jobenjada) [inline apps/web/pages/api/v1/environments/[environmentId]/surveys/[surveyId]/og/index.tsx:18]: This doesnt look like the mockup, will give it a try :)

**Review by jobenjada — CHANGES_REQUESTED** [review]

File: apps/web/app/s/[surveyId]/page.tsx (hunk 3 above)
Comment (anjy7) [reply apps/web/app/s/[surveyId]/page.tsx:58]: Sure, I thought the title here was the website name and the dynamic one is only inside the image, small change - will do it soon

File: apps/web/pages/api/v1/environments/[environmentId]/surveys/[surveyId]/og/index.tsx (hunk 4 above)
Comment (anjy7) [reply apps/web/pages/api/v1/environments/[environmentId]/surveys/[surveyId]/og/index.tsx:18]: Oh, can you please specify the changes?

File: apps/web/app/s/[surveyId]/layout.tsx (hunk 1 above)
Comment (anjy7) [reply apps/web/app/s/[surveyId]/layout.tsx:5]: Yup, it’s easily possible to change it to dynamic title. Will change it now

File: apps/web/app/s/[surveyId]/layout.tsx (hunk 2 above)
Comment (anjy7) [reply apps/web/app/s/[surveyId]/layout.tsx]: Screenshot from 2023-10-18 16-24-43
Done..

**Review by jobenjada — APPROVED** [review]

**Comment by jobenjada** [comment]
looks great, handing it over for final code review before merge 🤓

File: apps/web/pages/api/v1/environments/[environmentId]/surveys/[surveyId]/og/index.tsx
Comment (mattinannt) [inline apps/web/pages/api/v1/environments/[environmentId]/surveys/[surveyId]/og/index.tsx:1]: @anjy7 Can you please move this file to the app directory as we are currently deprecating the pages directory and making it ready to be deleted.
Check out [Route Handlers](https://nextjs.org/docs/app/building-your-application/routing/route-handlers) to see how the app directory is handling apis.

File: apps/web/package.json [hunk 5]
```diff
@@ -27,7 +27,9 @@
     "@radix-ui/react-dropdown-menu": "^2.0.6",
     "@react-email/components": "^0.0.7",
     "@sentry/nextjs": "^7.73.0",
+    "@vercel/og": "^0.5.17",
     "@t3-oss/env-nextjs": "^0.7.1",
+
```
Comment (mattinannt) [inline apps/web/package.json:32]: please remove the empty line

File: apps/web/pages/api/v1/environments/[environmentId]/surveys/[surveyId]/og/index.tsx
Comment (mattinannt) [reply apps/web/pages/api/v1/environments/[environmentId]/surveys/[surveyId]/og/index.tsx:1]: I would move the file to `api/v1/og` to make the url simpler

**Review by mattinannt — COMMENTED** [review]
@anjy7 thanks a lot for the PR, looks great! From the technical side I would like to see a few more things changed before we merge the changes.

File: apps/web/pages/api/v1/environments/[environmentId]/surveys/[surveyId]/og/index.tsx
Comment (anjy7) [reply apps/web/pages/api/v1/environments/[environmentId]/surveys/[surveyId]/og/index.tsx:1]: Sure, I’ll make the changes

**Comment by anjy7** [comment]
Hey @mattinannt, I have made the changes. Also the share button wasn't working on the survey as ig someone had put a check whether the survey is single use or not ( this prevented the share dropdown opening if the survey is not single use ). I have removed it, please check it once :)

**Comment by mattinannt** [comment]
@anjy7 thanks a lot for the changes and also fixing the bug on the way :-) 💪 I will merge the PR now :-)

**Review by mattinannt — APPROVED** [review]

**Comment by mattinannt** [comment]
@anjy7 Unfortunately there is still a build error in the PR:
```
@formbricks/web:build: app/api/v1/og/route.tsx
@formbricks/web:build: Type error: Route "app/api/v1/og/route.tsx" does not match the required types of a Next.js Route.
@formbricks/web:build:   Invalid configuration "fetchCache":
@formbricks/web:build:     Expected ""auto" | "force-cache" | "force-no-store" | "only-no-store" | "default-no-store" | "default-cache" | "only-cache" | undefined", got "false".
```

Can you please solve this? :-) Then we are ready to merge! :-)

**Comment by anjy7** [comment]
Hey @mattinannt, I’ll look into this in a bit

**Comment by anjy7** [comment]
Hi @mattinannt, I built it locally and all the test cases passed..
this was the error message in the above failed merge  - `  ../../packages/lib/storage/service.ts:70:26
  Type error: Property 'getType' does not exist on type 'typeof import("/home/runner/work/formbricks/formbricks/node_modules/@types/mime/index")'.` 
Although it shouldnt be there as the build is successful locally, could you please maybe try merging it again?

**Comment by anjy7** [comment]
should work now @mattinannt 
image

**Pull request merged**